File: mm2mw.pl
#########################################################################
# Filter()
#
# filters lines to correct syntax
#
# args: 1) infile - file to have words converted
# args: 2) outfile - file to output
#
#########################################################################
sub Filter {

	my $this = shift;
	my $args = {
		infile => " ",
		outfile => " ",
		@_,			# Override previous attributes
	};
	my $infile = $args->{infile};
	my $outfile = $args->{outfile};

        my @slashparts;
        my $filename;
	my @spaceparts;
        my $name;

        my $is_table = 0;

        open(INFILE, "<$infile");
        open(OUTFILE, ">$outfile");
        while (<INFILE>){

                ##########
                # line skips
                #
                next if /^----$/;       # remove unneeded header lines

                ##########
                # regexp substitutions
                #

                # list convert - 1st layer
                #
                s/^ \*(.*?)/\*$1/g;

                # list convert - 2nd layer
                #
                s/^  \*(.*?)/\*\*$1/g;

                # list convert - 3rd layer
                #
                s/^   \*(.*?)/\*\*\*$1/g;

                # list convert - 4th layer
                #
                s/^    \*(.*?)/\*\*\*\*$1/g;

                # list convert - 5th layer
                #
                s/^     \*(.*?)/\*\*\*\*\*$1/g;

                # list convert - numerical
                #
                s/^ 1.(.*?)/\#$1/g;

                # link convert - [[BR]] -> <br>
                #
                s/\[\[BR\]\]/\<br\>/g;

                # link convert superscripted - ^ * ^ -> <sup> * </sup>
                #
                s/\^(.*?)\^/\<sup\>$1\<\/sup\>/g;

                # link convert subscripted - ,, * ,, -> <sub> * </sub>
                #
                s/\,\,(.*?)\,\,/\<sub\>$1\<\/sub\>/g;

                # link convert smaller - ~- * -~ -> <small> * </small>
                #
                s/\~-(.*?)-\~/\<small\>$1\<\/small\>/g;

                # link convert larger - ~+ * +~ -> <big> * </big>
                #
                s/\~\+(.*?)\+\~/\<big\>$1\<\/big\>/g;

                # link convert - [# * ] -> [[ * ]]
                #
                s/\[\#(.*?)\]/\[\[\#$1\]\]/g;

                # link convert - [" * "] -> [[ * ]]
                #
                s/\[\"(.*?)\"\]/\[\[$1\]\]/g;

                # link convert - [: * ] -> [[ * ]]
                #
                s/\[\:(.*?)\]/\[\[$1\]\]/g;

                # link convert - __ * __ -> <u> * </u>
                #
                s/__(.*?)__/\<u\>$1\<\/u\>/g;

                # link convert - ` * ` -> <tt> * </tt>
                #
                s/`(.*?)`/\<tt\>$1\<\/tt\>/g;

                # link convert - {{{ * }}} -> <code><nowiki> * </code>
                # (if on same line)
                #
                s/\{\{\{(.*?)\}\}\}/<code\>\<nowiki\>$1\<\/nowiki\>\<\/code\>/g;

                # link convert - {{{ *  -> <pre> *
                # (}}} on different line)
                #
                s/\{\{\{(.*?)/\<pre\>\<nowiki\>$1/g;

                # link convert -  * }}} ->  * <\pre><\nowiki>
                # (if on same line)
                #
                s/(.*?)\}\}\}/$1\<\/nowiki\><\/pre\>/g;

                # strikethrough convert
                #
                s/--\((.*?)\)--/\<s\>$1\<\/s\>/g;

                # link convert - [wiki:cat * ] -> [[cat| * ]]
                #
                $this->ConvertWikiLinks(line=>\$_);

                # convert CamelCase links
                #
                $this->ConvertCamelLinks(line=>\$_);

                # link convert - /CommentPage -> Talk:PageName
                #
                @slashparts = split(/\//,$infile);
                my $filename = $slashparts[$#slashparts];
                @spaceparts = split(/(?=[A-Z])/,$filename);
                $name = join(" ",@spaceparts);
                s/\/CommentPage/\[\[Talk:$name\]\]/g;

                # Convert definition lists
                #
                # Partially built out, but not implemented
                #
                # MM has term, colon, colon, space, definition
                # MW has semicolon, space, term, space, colon, space, definition
                # s/(.*)\:\:\s(.*)/\;\s$1\s\:\s$2/g;

                ##########
                # deletions
                #

                # `` - don't need these (ex Myth``Blog on front page)
                #
                s/``//g;

                ##########
                # tables
                #

                # start of table/row?
                if(/^\|\|/) {
                        if($is_table == 0) {
                                # start of table
                                $is_table = 1;
                                s/^\|\|/\n\{\|border=1 cellpadding=1 cellspacing=0\|\n\|/;
                                }
                        else {
                                # start of row
                                s/^\|\|/\n\|-\n\|/;
                                }
                        }
                else {
                        # end of table?
                        # (i.e. in a table and the line doesn't start '||')
                        if ($is_table == 1) {
                                $is_table = 0;
                                s/(.*)$/$1\n\|\}\n/;
                                }
                        }

                # end of row (we just zap that)
                s/\|\|\s+$//;

                # all other '||' are just cell indicators
                s/\|\|/\n\|/g;

                print OUTFILE "$_";
        }

        # source text ended with unclosed table
        print OUTFILE "\n\|\}\n" if ($is_table == 1);

        close OUTFILE;
        close INFILE;
}

File: mm2mw.rb
#!/usr/bin/env ruby
# coding: utf-8

# MoinMoin to MediaWiki converter
# it is just the rewritten mm2mw.pl in ruby

input = ARGV[0]
output = ARGV[1]

output_array = []

def list_convert line,depth=5
  line = list_convert(line,depth-1) unless depth == 1
  line.gsub(%r{^#{' '*depth}\*(.*?)}){ "*"*depth + $1}
end

# link convert - [wiki:cat * ] -> [[cat| * ]]
# TODO: Camelcase links
# TODO: Link comments
def convert_wiki_links line
  line = line.gsub(/[\[]([^\s]+) (.*?)[\]]/){ "[[#{$1}| #{$2}]]" }
end

@@is_table = 0

open(input).readlines.each do |line|

  # line skips
  next if line =~ /^----$/;       # remove unneeded header lines

  # regexp substitutions

  #line = replace_hun_chars line

  # list convert
  line = list_convert(line)

  # list convert - numerical
  line = line.gsub(/^ 1.(.*?)/){ "##{$1}" }

  # link convert - [[BR]] -> <br>
  line = line.gsub(/\[\[BR\]\]/, "<br>" )

  # link convert superscripted - ^ * ^ -> <sup> * </sup>
  line = line.gsub(/\^(.*?)\^/){ "<sup>#{$1}</sup>" }

  # link convert subscripted - ,, * ,, -> <sub> * </sub>
  line = line.gsub(/\,\,(.*?)\,\,/){ "<sub>#{$1}</sub>" }

  # link convert smaller - ~- * -~ -> <small> * </small>
  line = line.gsub(/\~-(.*?)-\~/){ "<small>#{$1}</small>" }

  # link convert larger - ~+ * +~ -> <big> * </big>
  line = line.gsub(/\~\+(.*?)\+\~/){ "<big>#{$1}</big>" }

  # link convert - [# * ] -> [[ * ]]
  line = line.gsub(/\[\#(.*?)\]/){ "[[#{$1}]]" }

  # link convert - [" * "] -> [[ * ]]
  line = line.gsub(/\[\"(.*?)\"\]/){ "[[#{$1}]]" }

  # link convert - [: * ] -> [[ * ]]
  line = line.gsub(/\[\:(.*?)\]/){"[[#{$1}]]" }

  # link convert - __ * __ -> <u> * </u>
  line = line.gsub(/__(.*?)__/){ "<u>#{$1}</u>" }

  # link convert - ` * ` -> <tt> * </tt>
  line = line.gsub(/`(.*?)`/){ "<tt>#{$1}</tt>" }

  # link convert - {{{ * }}} -> <code><nowiki> * </code>
  # (if on same line)
  line = line.gsub(/\{\{\{(.*?)\}\}\}/){ "<code><nowiki>#{$1}</nowiki></code>" }

  # link convert - {{{ *  -> <pre> *
  # (}}} on different line)
  line = line.gsub(/\{\{\{(.*?)/){ "<pre><nowiki>#{$1}" }

  # link convert -  * }}} ->  * <\pre><\nowiki>
  # (if on same line)
  line = line.gsub(/(.*?)\}\}\}/){ "#{$1}</nowiki></pre>" }

  # strikethrough convert
  line = line.gsub(/--\((.*?)\)--/){ "<s>#{$1}</s>" }

  # link convert - [wiki:cat * ] -> [[cat| * ]]
  line = convert_wiki_links( line )
        
  # don't need this
  line =line.gsub('``','')


  # tables

  # start of table/row?
  if line =~ /^\|\|/
    if @@is_table == 0
      # start of table
      @@is_table = 1;
      line = line.gsub(/^\|\|/, "\n{|border=1 cellpadding=1 cellspacing=0|\n|")
    else
      # start of row
      line = line.gsub(/^\|\|/, "\n|-\n|")
    end
  else
    # end of table?
    # (i.e. in a table and the line doesn't start '||')
    if @@is_table == 1
      @@is_table = 0
      line = line.gsub(/(.*)$/){ $1 +"\n|}\n"}
    end
  end

  # end of row (we just zap that)
  line = line.gsub(/\|\|\s+$/,'')

  # all other '||' are just cell indicators
  line = line.gsub(/\|\|/,"\n|")

  output_array << line

end

# Write output
File.open(output, "w") do |f|
  output_array.each{|l| f.puts l}
end
